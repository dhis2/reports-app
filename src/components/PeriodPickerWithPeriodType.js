import i18n from '@dhis2/d2-i18n'
import { DropDown, PeriodPicker } from '@dhis2/d2-ui-core'
import parsePeriod from 'd2/period/parser'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { selectPeriod, selectPeriodType } from '../redux/actions/reportPeriod'
import { getTranslatedPeriodTypes } from '../redux/selectors/reportPeriod/getTranslatedPeriodTypes'
import { formLabel } from '../utils/styles/shared.js'

const periodTypeLabelText = i18n.t('Select Period Type')

export function PeriodPickerWithPeriodType({
    selectPeriodType,
    selectPeriod,
    label,
    collection,
    selectedPeriodType,
    selectedPeriod,
}) {
    return (
        <div>
            <span className={formLabel.className}>{label}</span>
            <DropDown
                fullWidth
                emptyLabel={periodTypeLabelText}
                hintText={periodTypeLabelText}
                menuItems={collection}
                onChange={selectPeriodType}
                value={selectedPeriodType}
            />
            {selectedPeriodType && (
                <PeriodPicker
                    periodType={selectedPeriodType}
                    onPickPeriod={selectPeriod}
                />
            )}
            {selectedPeriod && (
                <div className="parsed-period">
                    {parsePeriod(selectedPeriod).name}
                </div>
            )}
            <style jsx>{`
                .parsed-period {
                    margin-bottom: 16px;
                }
            `}</style>
            {formLabel.styles}
        </div>
    )
}

PeriodPickerWithPeriodType.propTypes = {
    selectPeriodType: PropTypes.func.isRequired,
    selectPeriod: PropTypes.func.isRequired,
    label: PropTypes.string.isRequired,
    collection: PropTypes.array.isRequired,
    selectedPeriodType: PropTypes.string,
    selectedPeriod: PropTypes.string,
}

const mapStateToProps = state => ({
    ...state.reportPeriod,
    collection: getTranslatedPeriodTypes(state),
})

export default connect(
    mapStateToProps,
    {
        selectPeriodType,
        selectPeriod,
    }
)(PeriodPickerWithPeriodType)
