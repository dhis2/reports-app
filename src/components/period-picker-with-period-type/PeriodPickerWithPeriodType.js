/* React */
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'

/* d2-ui components */
import { DropDown, PeriodPicker } from '@dhis2/d2-ui-core'

/* App context */
import AppContext from '../../pages/AppContext'

/* Actions */
import loadPeriodTypes from '../../redux/actions/periodTypes'
import pluckPeriodTypes from '../../redux/selectors/periodTypes'

/* i18n */

/* styles */
import styles from '../../utils/styles'

export class PeriodPickerWithPeriodType extends PureComponent {
    state = {
        periodTypes: [],
        selectedPeriodType: null,
        selected: null,
    }

    /* FIXME: right now d2-ui periodPicker forces us to pass d2 through old  context api */
    getChildContext() {
        return {
            d2: this.props.d2,
        }
    }

    componentDidMount() {
        const { loadPeriodTypes, periodTypes } = this.props
        if (Array.isArray(periodTypes) && periodTypes.length === 0) {
            loadPeriodTypes()
        }
    }

    onChangePeriodType = event => {
        const selectedPeriodType = event.target.value
        this.setState({
            selectedPeriodType,
        })
    }

    renderPeriodTypeDropdown = () => {
        const msg = i18n.t('Select Period Type')
        return (
            <DropDown
                value={this.state.selectedPeriodType}
                onChange={this.onChangePeriodType}
                menuItems={this.props.periodTypes}
                fullWidth
                emptyLabel={msg}
                hintText={msg}
            />
        )
    }

    render() {
        const { periodTypes } = this.props
        const error =
            periodTypes instanceof Error ? periodTypes.toString() : null

        return (
            <div>
                <span style={{ ...styles.formLabel, display: 'block' }}>
                    {this.props.label}
                </span>
                {error && <span style={styles.error}>{error}</span>}
                {!error && this.renderPeriodTypeDropdown()}
                {this.state.selectedPeriodType && (
                    <PeriodPicker
                        periodType={this.state.selectedPeriodType}
                        onPickPeriod={this.props.onChange}
                    />
                )}
            </div>
        )
    }
}

PeriodPickerWithPeriodType.propTypes = {
    d2: PropTypes.object.isRequired,
    onChange: PropTypes.func,
    label: PropTypes.string,
    loadPeriodTypes: PropTypes.func.isRequired,
    periodTypes: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
        .isRequired,
}

PeriodPickerWithPeriodType.defaultProps = {
    onChange: () => {},
    label: i18n.t('Period'),
}

/* FIXME: right now d2-ui periodPicker forces us to pass d2 through old  context api */
PeriodPickerWithPeriodType.childContextTypes = {
    d2: PropTypes.object.isRequired,
}

const PeriodPickerWithPeriodTypeWithContext = props => (
    <AppContext.Consumer>
        {appContext => (
            <PeriodPickerWithPeriodType d2={appContext.d2} {...props} />
        )}
    </AppContext.Consumer>
)

const mapStateToProps = ({ periodTypes }) => ({
    periodTypes: pluckPeriodTypes(periodTypes),
})

export default connect(
    mapStateToProps,
    {
        loadPeriodTypes,
    }
)(PeriodPickerWithPeriodTypeWithContext)
