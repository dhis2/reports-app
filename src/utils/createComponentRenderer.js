import React from 'react'
import { shallow } from 'enzyme'

export default function createComponentRenderer(Component, defaultProps) {
    return function(customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
