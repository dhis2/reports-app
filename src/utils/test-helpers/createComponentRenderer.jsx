import { shallow } from 'enzyme'
import React from 'react'

export default function createComponentRenderer(Component, defaultProps) {
    return function (customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
