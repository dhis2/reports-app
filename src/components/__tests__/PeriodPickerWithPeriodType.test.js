import createComponentRenderer from '../../utils/test-helpers/createComponentRenderer'
import PeriodPickerWithPeriodType from '../PeriodPickerWithPeriodType'

jest.mock('@dhis2/d2-ui-core', () => ({
    DropDown: 'DropDown',
    PeriodPicker: 'PeriodPicker',
}))

describe('Test <PeriodPickerWithPeriodType /> rendering:', () => {
    const selectPeriodTypeMock = jest.fn()
    const selectPeriodMock = jest.fn()
    const collectionMock = [
        { id: '1', displayName: 'Yes' },
        { id: '2', displayName: 'No' },
    ]

    const defaultProps = {
        selectPeriodType: selectPeriodTypeMock,
        selectPeriod: selectPeriodMock,
        label: 'This is my label',
        collection: collectionMock,
        ready: true,
        loadingError: '',
        selectedPeriodType: '',
    }

    const componentRenderer = createComponentRenderer(
        PeriodPickerWithPeriodType,
        defaultProps
    )

    afterEach(() => {
        jest.clearAllMocks()
    })

    it('It should render a <PeriodTypeDropDown/> when selectedPeriodType is falsy', () => {
        expect(componentRenderer()).toMatchSnapshot()
    })

    it('It should render a <PeriodTypeDropDown/> and <PeriodPicker/> when selectedPeriodType is truthy', () => {
        expect(
            componentRenderer({ selectedPeriodType: 'Quarterly' })
        ).toMatchSnapshot()
    })
})
