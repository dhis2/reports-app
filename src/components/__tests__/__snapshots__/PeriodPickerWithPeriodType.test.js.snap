// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <PeriodPickerWithPeriodType /> rendering: It should render a <PeriodTypeDropDown/> and <PeriodPicker/> when selectedPeriodType is truthy 1`] = `
<div>
  <span>
    This is my label
  </span>
  <DropDown
    emptyLabel="Select Period Type"
    fullWidth={true}
    hintText="Select Period Type"
    menuItems={
      Array [
        Object {
          "displayName": "Yes",
          "id": "1",
        },
        Object {
          "displayName": "No",
          "id": "2",
        },
      ]
    }
    onChange={[Function]}
    value="Quarterly"
  />
  <PeriodPicker
    onPickPeriod={[Function]}
    periodType="Quarterly"
  />
  <style>
    
                    .parsed-period {
                        margin-bottom: 16px;
                    }
                
  </style>
</div>
`;

exports[`Test <PeriodPickerWithPeriodType /> rendering: It should render a <PeriodTypeDropDown/> when selectedPeriodType is falsy 1`] = `
<div>
  <span>
    This is my label
  </span>
  <DropDown
    emptyLabel="Select Period Type"
    fullWidth={true}
    hintText="Select Period Type"
    menuItems={
      Array [
        Object {
          "displayName": "Yes",
          "id": "1",
        },
        Object {
          "displayName": "No",
          "id": "2",
        },
      ]
    }
    onChange={[Function]}
    value=""
  />
  <style>
    
                    .parsed-period {
                        margin-bottom: 16px;
                    }
                
  </style>
</div>
`;

exports[`Test <PeriodPickerWithPeriodType /> rendering: It should render the parsed period when a period has been selected 1`] = `
<div>
  <span>
    This is my label
  </span>
  <DropDown
    emptyLabel="Select Period Type"
    fullWidth={true}
    hintText="Select Period Type"
    menuItems={
      Array [
        Object {
          "displayName": "Yes",
          "id": "1",
        },
        Object {
          "displayName": "No",
          "id": "2",
        },
      ]
    }
    onChange={[Function]}
    value="Daily"
  />
  <PeriodPicker
    onPickPeriod={[Function]}
    periodType="Daily"
  />
  <div
    className="parsed-period"
  >
    January 1, 2018
  </div>
  <style>
    
                    .parsed-period {
                        margin-bottom: 16px;
                    }
                
  </style>
</div>
`;
